presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java

```diff
@@ -1296,6 +1309,10 @@ public PlanNode visitTableScan(TableScanNode node, RewriteContext<Void> context)
                 return node;
             }
 
+            if (isPreferLocalUnion(session) && fragmentPartitioningHandle.equals(SOURCE_DISTRIBUTION)) {
```

Comment:
Also for case of
`bucketed table with bucketed distribution UNION ALL table with system paritioing (source distribution)`, source distribution takes priority